src/object_store/src/object/s3.rs

```diff
@@ -12,18 +12,266 @@
+
+        let mut uploaded_parts = Vec::with_capacity(join_handles.len());
+        for result in try_join_all(join_handles)
+            .await
+            .map_err(ObjectError::internal)?
+        {
+            uploaded_parts.push(result?);
+        }
+
+        let completed_parts = Some(
+            uploaded_parts
+                .iter()
+                .map(|(part_id, output)| {
+                    CompletedPart::builder()
+                        .set_e_tag(output.e_tag.clone())
+                        .set_part_number(Some(*part_id))
+                        .build()
+                })
+                .collect_vec(),
+        );
+
+        self.client
+            .complete_multipart_upload()
+            .bucket(&self.bucket)
+            .key(&self.key)
+            .upload_id(&self.upload_id)
+            .multipart_upload(
+                CompletedMultipartUpload::builder()
+                    .set_parts(completed_parts)
+                    .build(),
+            )
+            .send()
+            .await?;
+
+        Ok(())
+    }
+
+    async fn abort(&self) -> ObjectResult<()> {
+        // If any part uploads are currently in progress, those part uploads might or might
+        // not succeed. As a result, it might be necessary to abort a given multipart upload
+        // multiple times in order to completely free all storage consumed by all parts.
+        //
+        // To verify that all parts have been removed, so you don't get charged for the
+        // part storage, you should call the ListParts action and ensure that the parts list is
+        // empty.
+        //
+        // Reference: <https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html>
+        self.client
+            .abort_multipart_upload()
+            .bucket(&self.bucket)
+            .key(&self.key)
+            .upload_id(&self.upload_id)
+            .send()
+            .await?;
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl StreamingUploader for S3StreamingUploader {
+    fn write_bytes(&mut self, data: Bytes) -> ObjectResult<()> {
+        fail_point!("s3_write_bytes_err", |_| Err(ObjectError::internal(
+            "s3 write bytes error"
+        )));
+        let data_len = data.len();
+        self.not_uploaded_len += data_len;
+        self.buf.push(data);
+
+        if self.not_uploaded_len > self.part_size {
+            if self.part_end == 0 {
+                // Mark current slice of buffer to be the next part to be uploaded.
+                self.part_end = self.buf.len();
+            } else {
+                // `data` should be uploaded in the next part.
+                self.next_part_len += data_len;
+            }
+        }
+        if self.next_part_len >= MIN_PART_SIZE {
+            // Take a 16MiB part and upload it. `Bytes` performs shallow clone.
+            let part = self.buf.drain(..self.part_end).collect();
```

Comment:
If the input of `upload_part` becomes `impl IntoIter<Item = Bytes>`, we don't need an extra `collect` here.